design/custom-request-timeout.md

```diff
@@ -0,0 +1,256 @@
+              idle: 2s
+          retryPolicy:
+              # the number of retries it should make on failure
+              maxRetries: "7"
+              # the conditions for which it should reattempt
+              onStatusCodes:
+              - 501
+              - 502
+              # timeout between the requests, if greater than request timeout, this field is ignored
+              perTryTimeout: 75ms
+        - match: /static-site
+          timeoutPolicy:
+              request: 200ms
+              idle: 1s
+          retryPolicy:
+              maxRetries: 2
+              # for all statusCodes but 504, retry the request
+              onStatusCodes:
+              - 5xx
+              perTryTimeout: 75ms
+```
+
+The naming conventions of the proposed YAML fields are inspired from [HTTP Timeouts](https://tools.ietf.org/id/draft-thomson-hybi-http-timeout-00.html#rfc.section.4).
+
+# Detailed Design
+
+## Reading from YAML
+The ingressroute spec struct will be updated to contain Timeout and Retry struct members
+
+
+```golang
+// contour/apis/contour/v1beta1/ingressroute.go
+
+// Route contains the set of routes for a virtual host
+type Route struct {
+    [... other members ...]
+	// The request timeout for this route
+	Timeout TimeoutPolicy `json:"timeoutPolicy,omitempty"`
+	// The retry attempts for this route
+	Retry RetryPolicy `json:"retryPolicy,omitempty"`
+}
+
+// TimeoutPolicy defines the attributes associated with timeout
+type TimeoutPolicy struct {
+	// Timeout for receiving a response from the server after processing a request from client in milliseconds
+	Request	string `json:"request"`
+	// Timeout for an idle connection to terminate when there are no active requests
+	Idle string `json:"idle"`
+}
+
+// RetryPolicy defines the attributes associated with retrying policy
+type RetryPolicy struct {
+	// MaxRetries is maximum allowed number of retries
+	MaxRetries	int `json:"maxRetries"`
+	// Perform retry on failed requests with the matched status codes or aggregated as 5xx
+	OnStatusCodes []string `json:"onStatusCodes"`
+	// PerTryTimeout specifies the timeout per retry attempt. Ignored if OnStatusCodes are empty
+	PerTryTimeout string `json:"perTryTimeout"`
+}
+```
+
+## Intermediate form in the DAG
+We store unmarshalled structs from ingress' and ingressroute's yaml into a common dag struct route for both. 
+There are three approaches to do change:
+
+### Approach 1
+Just add IdleTimeout as another member in the struct as below as Idle Timeout is the only new member
+in the _current_ proposed CRD.
+
+```golang
+// Approach 1: contour/internal/dag/dag.go
+
+type Route struct {
+    [... other members ...]
+    IdleTimeout time.Duration
+}
+
+```
+
+### Approach 2
+
+Consolidate timeout and retry members of dag's Route struct to their own respective struct.
+
+```golang
+// Approach 2: contour/internal/dag/dag.go
+
+type RouteTimeout struct {
+    Request time.Duration
+    Idle    time.Duration
```

Comment:
Please have a look at the discussions when this was added to Ingress. There is some messiness about dealing with 'Infinite' timeout, which envoy interprets as 0, and 'not set' which is _not_ the zero value of this field.

Sadly this grossness will have to be replicated here because people _will_ ask for infinite timeout.